File: facecoolalert/src/main/java/com/facecoolalert/utils/FileUtils.java
package com.facecoolalert.utils;

import android.content.ContentUris;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.provider.DocumentsContract;
import android.provider.MediaStore;
import android.util.Log;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.zip.ZipInputStream;

public class FileUtils {

    public static String getFriendlyPathFromUri(Context context, Uri uri) {
        String path = null;
        Log.d("FileUtils", "URI: " + uri+" Path : "+uri.getPath()+" raw "+uri.toString()+" ");

        // Check for "file" scheme (local file)
        if ("file".equalsIgnoreCase(uri.getScheme())) {
            path = uri.getPath();
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT && DocumentsContract.isDocumentUri(context, uri)) {
            // Handle document URI (e.g., Google Drive, Downloads)
            String documentId = DocumentsContract.getDocumentId(uri);
            String authority = uri.getAuthority();

            if ("com.android.providers.downloads.documents".equals(authority)) {
                Uri contentUri = ContentUris.withAppendedId(Uri.parse("content://downloads/public_downloads"), Long.valueOf(documentId));
                path = getDataColumn(context, contentUri, null, null);
            } else if ("com.android.providers.media.documents".equals(authority)) {
                String[] split = documentId.split(":");
                String type = split[0];

                Uri contentUri = null;
                if ("image".equals(type)) {
                    contentUri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
                } else if ("video".equals(type)) {
                    contentUri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI;
                } else if ("audio".equals(type)) {
                    contentUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
                }

                String selection = MediaStore.Images.Media._ID + "=?";
                String[] selectionArgs = new String[]{split[1]};
                path = getDataColumn(context, contentUri, selection, selectionArgs);
            }
        } else {
            // Handle other URIs (e.g., "content" scheme)
            String[] projection = {MediaStore.Images.Media.DATA};
            Cursor cursor = context.getContentResolver().query(uri, projection, null, null, null);

            if (cursor != null) {
                int column_index = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA);
                if (cursor.moveToFirst()) {
                    path = cursor.getString(column_index);
                }
                cursor.close();
            }
        }

        // Extract the friendly path from the absolute path
        if (path != null) {
            String[] segments = path.split("/");
            if (segments.length > 1) {
                return segments[segments.length - 2] + "/" + segments[segments.length - 1];
            }
        }

        if(path==null)
            path=getFileNameFromUri(context,uri);

        return path;
    }

    private static String getDataColumn(Context context, Uri uri, String selection, String[] selectionArgs) {
        Cursor cursor = null;
        String column = MediaStore.Images.Media.DATA;
        String[] projection = {column};

        try {
            cursor = context.getContentResolver().query(uri, projection, selection, selectionArgs, null);
            if (cursor != null && cursor.moveToFirst()) {
                int column_index = cursor.getColumnIndexOrThrow(column);
                return cursor.getString(column_index);
            }
        } catch (Exception e) {
            Log.e("FileUtils", "Error getting data column for uri: " + uri, e);
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }

        return null;
    }

    public static String getFileNameFromUri(Context context, Uri uri) {
        String fileName = uri.getPath().toString();
        Cursor cursor = null;

        try {
            String[] projection = {MediaStore.Images.Media.DISPLAY_NAME};
            cursor = context.getContentResolver().query(uri, projection, null, null, null);

            if (cursor != null && cursor.moveToFirst()) {
                int columnIndex = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DISPLAY_NAME);
                fileName = cursor.getString(columnIndex);
            }
        } catch (Exception e) {
            // Handle exceptions, log, or perform error handling as needed
            e.printStackTrace();
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }

        return fileName;
    }


    public static byte[] readAllBytes(InputStream inputStream) throws IOException {
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        byte[] buffer = new byte[1024];
        int len;
        while ((len = inputStream.read(buffer)) > 0) {
            byteArrayOutputStream.write(buffer, 0, len);
        }
        return byteArrayOutputStream.toByteArray();
    }


}

File: facecool/src/main/java/com/facecool/utils/FileUtils.kt
package com.facecool.utils

import android.content.Context
import android.content.Intent
import android.database.Cursor
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.provider.DocumentsContract
import android.provider.MediaStore
import androidx.core.content.FileProvider
import com.facecool.BuildConfig
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.File
import java.io.FileReader
import java.io.FileWriter
import java.io.IOException


fun getExternalFilesDirCompat(folderName: String, context: Context): File? {
    var file: File? = null
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
        file = context.getExternalFilesDir(folderName)
    } else {
        val externalDir = Environment.getExternalStorageDirectory()
        if (externalDir != null) {
            file =
                File(externalDir.absolutePath + "/Android/data/" + context.packageName + "/files/" + folderName)
            if (!file.exists()) {
                file.mkdirs()
            }
        }
    }
    return file
}


fun Context.shareReport(cvsData: String) {
    try {
        val filename = "student_report_for_" + System.currentTimeMillis()
            .getReadableDate(pattern = "MMMM_dd_yyyy") + ".csv"
        val file =
            File(getExternalFilesDirCompat("reports", this), filename)
        val fw = FileWriter(file.absoluteFile)
        val bw = BufferedWriter(fw)
        bw.write(cvsData)
        bw.close()
        if (file.exists()) {
            val uri = FileProvider.getUriForFile(
                this,
                BuildConfig.APPLICATION_ID + ".provider",
                file
            )
            val intent = Intent(Intent.ACTION_SEND)
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            intent.type = "*/*"
            intent.putExtra(Intent.EXTRA_STREAM, uri)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            startActivity(intent)
        }
    } catch (e: IOException) {
        e.printStackTrace()
    }
}

fun Context.shareData(data: String) {
    try {
        val filename = "face_cool_db_" + System.currentTimeMillis().getReadableDate(pattern = "MMMM_dd_yyyy_HH_mm") + ".fcud"
        val file =
            File(getExternalFilesDirCompat("student_data", this), filename)
        val fw = FileWriter(file.absoluteFile)
        val bw = BufferedWriter(fw)
        bw.write(data)
        bw.close()
        if (file.exists()) {
            val uri = FileProvider.getUriForFile(
                this,
                BuildConfig.APPLICATION_ID + ".provider",
                file
            )
            val intent = Intent(Intent.ACTION_SEND)
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            intent.type = "*/*"
            intent.putExtra(Intent.EXTRA_STREAM, uri)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            startActivity(intent)
        }
    } catch (e: IOException) {
        e.printStackTrace()
    }
}

fun loadData(file: File): String {

    //Read text from file
    val text = StringBuilder()

    try {
        val br = BufferedReader(FileReader(file))
        var line: String?
        while (br.readLine().also { line = it } != null) {
            text.append(line)
            text.append('\n')
        }
        br.close()
    } catch (e: IOException) {
        //You'll need to add proper error handling here
    }

    return text.toString()

}


//fun getPath(context: Context?, uri: Uri): String? {
//
//    // check here to KITKAT or new version
//    val isKitKat = Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT
//
//    // DocumentProvider
//    if (isKitKat && DocumentsContract.isDocumentUri(context, uri)) {
//
//        // ExternalStorageProvider
//        if (isExternalStorageDocument(uri)) {
//            val docId = DocumentsContract.getDocumentId(uri)
//            val split = docId.split(":".toRegex()).dropLastWhile { it.isEmpty() }
//                .toTypedArray()
//            val type = split[0]
//            if ("primary".equals(type, ignoreCase = true)) {
//                return (Environment.getExternalStorageDirectory().toString() + "/"
//                        + split[1])
//            }
//        } else if (isDownloadsDocument(uri)) {
//            val id = DocumentsContract.getDocumentId(uri)
//            val contentUri = ContentUris.withAppendedId(
//                Uri.parse("content://downloads/public_downloads"),
//                java.lang.Long.valueOf(id)
//            )
//            return getDataColumn(context, contentUri, null, null)
//        } else if (isMediaDocument(uri)) {
//            val docId = DocumentsContract.getDocumentId(uri)
//            val split = docId.split(":".toRegex()).dropLastWhile { it.isEmpty() }
//                .toTypedArray()
//            val type = split[0]
//            var contentUri: Uri? = null
//            if ("image" == type) {
//                contentUri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI
//            } else if ("video" == type) {
//                contentUri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI
//            } else if ("audio" == type) {
//                contentUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
//            }
//            val selection = "_id=?"
//            val selectionArgs = arrayOf(split[1])
//            return getDataColumn(
//                context, contentUri, selection,
//                selectionArgs
//            )
//        }
//    } else if ("content".equals(uri.scheme, ignoreCase = true)) {
//
//        // Return the remote address
//        return if (isGooglePhotosUri(uri)) uri.lastPathSegment else getDataColumn(
//            context,
//            uri,
//            null,
//            null
//        )
//    } else if ("file".equals(uri.scheme, ignoreCase = true)) {
//        return uri.path
//    }
//    return null
//}


object FilePath {
    /**
     * Method for return file path of Gallery image/ Document / Video / Audio
     *
     * @param context
     * @param uri
     * @return path of the selected image file from gallery
     */
    fun getPath(context: Context, uri: Uri): String? {

        // check here to KITKAT or new version
        val isKitKat = Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT

        // DocumentProvider
        if (isKitKat && DocumentsContract.isDocumentUri(context, uri)) {

            // ExternalStorageProvider
            if (isExternalStorageDocument(uri)) {
                val docId = DocumentsContract.getDocumentId(uri)
                val split = docId.split(":".toRegex()).dropLastWhile { it.isEmpty() }
                    .toTypedArray()
                val type = split[0]
                if ("primary".equals(type, ignoreCase = true)) {
                    return (Environment.getExternalStorageDirectory().toString() + "/"
                            + split[1])
                }
            } else if (isDownloadsDocument(uri)) {
                val id = DocumentsContract.getDocumentId(uri)
//                val contentUri = ContentUris.withAppendedId(
////                    Uri.parse("content://downloads/public_downloads"),
//                    Uri.parse("content://com.android.providers.downloads.documents/"),
//                    java.lang.Long.valueOf(id)
//                )
//                return getDataColumn(context, contentUri, null, null)
                return getDataColumn(context, uri, null, null)
            } else if (isMediaDocument(uri)) {
                val docId = DocumentsContract.getDocumentId(uri)
                val split = docId.split(":".toRegex()).dropLastWhile { it.isEmpty() }
                    .toTypedArray()
                val type = split[0]
                var contentUri: Uri? = null
                if ("image" == type) {
                    contentUri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI
                } else if ("video" == type) {
                    contentUri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI
                } else if ("audio" == type) {
                    contentUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
                }
                val selection = "_id=?"
                val selectionArgs = arrayOf(split[1])
                return contentUri?.let { getDataColumn(context, it, selection, selectionArgs) }
            }
        } else if ("content".equals(uri.scheme, ignoreCase = true)) {

            // Return the remote address
            return if (isGooglePhotosUri(uri)) uri.lastPathSegment else getDataColumn(
                context,
                uri,
                null,
                null
            )
        } else if ("file".equals(uri.scheme, ignoreCase = true)) {
            return uri.path
        }
        return null
    }

    /**
     * Get the value of the data column for this Uri. This is useful for
     * MediaStore Uris, and other file-based ContentProviders.
     *
     * @param context
     * The context.
     * @param uri
     * The Uri to query.
     * @param selection
     * (Optional) Filter used in the query.
     * @param selectionArgs
     * (Optional) Selection arguments used in the query.
     * @return The value of the _data column, which is typically a file path.
     */
    fun getDataColumn(context: Context, uri: Uri, selection: String?, selectionArgs: Array<String>?): String? {
        var cursor: Cursor? = null
        val column = "_data"
        val projection = arrayOf(column)
        try {
            cursor = context.contentResolver.query(uri, projection, selection, selectionArgs, null)
            if (cursor != null && cursor.moveToFirst()) {
                val index = cursor.getColumnIndexOrThrow(column)
                return cursor.getString(index)
            }
        } finally {
            cursor?.close()
        }
        return null
    }

    /**
     * @param uri
     * The Uri to check.
     * @return Whether the Uri authority is ExternalStorageProvider.
     */
    fun isExternalStorageDocument(uri: Uri): Boolean {
        return "com.android.externalstorage.documents" == uri
            .authority
    }

    /**
     * @param uri
     * The Uri to check.
     * @return Whether the Uri authority is DownloadsProvider.
     */
    fun isDownloadsDocument(uri: Uri): Boolean {
        return "com.android.providers.downloads.documents" == uri
            .authority
    }

    /**
     * @param uri
     * The Uri to check.
     * @return Whether the Uri authority is MediaProvider.
     */
    fun isMediaDocument(uri: Uri): Boolean {
        return "com.android.providers.media.documents" == uri
            .authority
    }

    /**
     * @param uri
     * The Uri to check.
     * @return Whether the Uri authority is Google Photos.
     */
    fun isGooglePhotosUri(uri: Uri): Boolean {
        return "com.google.android.apps.photos.content" == uri
            .authority
    }
}

